### PR TITLE
Normalize forest sample weights

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

Fixes test deselected in https://github.com/uxlfoundation/scikit-learn-intelex/pull/3231
Accompanies https://github.com/uxlfoundation/scikit-learn-intelex/pull/3292
Combined CI: http://intel-ci.intel.com/f170e0af-9e7a-f1a7-82dc-d4f5ef20c6a0

---

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with updates and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least a summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance and/or quality metrics have changed or why changes are not expected.
- [ ] I have extended the benchmarking suite and provided a corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

</details>

### CODE REVIEW

File: cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i [hunk 1]
```diff
@@ -1218,7 +1218,7 @@ bool OrderedRespHelperRandom<algorithmFPType, cpu>::findBestSplitOrderedFeature(
     ImpurityData left;
     ImpurityData right;
     IndexType iBest = -1;
-    algorithmFPType vBest;
+    intermSummFPType vBest;
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i:1221]: If you're going to do this, then it makes more sense to make `normalizedWeights` of `intermSummFPType`. It also would need to be modified in a lot of other places to be consistent.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 2]
```diff
@@ -47,6 +48,45 @@ namespace training
...
+    const size_t nRows = weights->getNumberOfRows();
+    if (!nRows) return empty;
+
+    ReadRows<algorithmFPType, cpu> srcBlock(const_cast<NumericTable *>(weights), 0, nRows);
+    s |= srcBlock.status();
+    if (!s) return empty;
+    const algorithmFPType * src = srcBlock.get();
+
+    algorithmFPType maxWeight = 0;
+    for (size_t i = 0; i < nRows; ++i)
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:70]: Could put vectorization pragma with reductor.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 3]
```diff
@@ -47,6 +48,45 @@ namespace training
...
+    {
+        if (src[i] > maxWeight) maxWeight = src[i];
+    }
+    if (!(maxWeight > 0)) return empty;
+
+    services::SharedPtr<HomogenNumericTableCPU<algorithmFPType, cpu> > normalized =
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:76]: Does this need to be a oneDAL table? It's just a one-dimensional array which could be passed as a unique or shared pointer.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 4]
```diff
@@ -47,6 +48,45 @@ namespace training
...
+        HomogenNumericTableCPU<algorithmFPType, cpu>::create(1, nRows, &s);
+    if (!s) return empty;
+
+    WriteOnlyRows<algorithmFPType, cpu> dstBlock(normalized.get(), 0, nRows);
+    s |= dstBlock.status();
+    if (!s) return empty;
+    algorithmFPType * dst = dstBlock.get();
+
+    for (size_t i = 0; i < nRows; ++i)
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: It could use ?rscl (preferrably) or ?scal from MKL.

Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: Updated

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 2 above)
Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:70]: Updated

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 3 above)
Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:76]: Downstream functionality (compute, computeForSpecificHelper, TrainBatchTask) all consume NumericTables so it sets up well for those

File: cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i (hunk 1 above)
Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i:1221]: The normalized table can't be intermSummFPType - it's read back as algorithmFPType by _helper.init/ReadRows, so a float32 fit needs a float32 table

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 4 above)
Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: On a second look, if it's going to copy beforehand, maybe it'd make more sense to replace with a vectorized loop. Could also set a pragma for the alignment, given that the new array is allocated through oneDAL.

File: cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i (hunk 1 above)
Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i:1221]: Then please change the dtype in all other places that calculate similar quantities to be consistent.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 4 above)
Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: I believe your MKL suggestion would be preferable

File: cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i (hunk 1 above)
Comment (ethanglaser) [reply cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i:1221]: I made one other change that I think is related to what you are eluding to, otherwise you're going to need to be more specific as I am not very familiar with these files.

**Comment by ethanglaser** [comment]
Updated combined CI: http://intel-ci.intel.com/f171b38c-632b-f18e-89b1-d4f5ef20c6a0

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i (hunk 4 above)
Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i]: I'm pretty sure a single loop doing a single simd store would be faster.

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 5]
```diff
@@ -47,6 +49,48 @@ namespace training
...
+    s |= dstBlock.status();
+    if (!s) return empty;
+    algorithmFPType * dst = dstBlock.get();
+
+    // Copy then divide in place via rscl (correctly-rounded division, also scale-invariant).
+    // The single-threaded xxrscl avoids nested threading inside the parallel tree build.
+    const DAAL_INT n    = static_cast<DAAL_INT>(nRows);
+    const DAAL_INT incx = 1;
+    services::internal::tmemcpy<algorithmFPType, cpu>(dst, src, nRows);
+    daal::internal::LapackInst<algorithmFPType, cpu>::xxrscl(&n, &maxWeight, dst, &incx);
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:90]: Maybe I do not understand something, but why does it normalize to a maximal weight, and not to a sum of weights?

Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:90]: If it uses the maximal weight, then all standardized row weights will be between 0 and 1, and the calculations would be more in line with unweighted routes.

Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:90]: Here is the scikit-learn implementation:
`normalized_sample_weight = sample_weight / np.sum(sample_weight)`
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/ensemble/_forest.py#L105
Maybe it's worth to make similar changes?

Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:90]: @Vika-F They are using those sample weights differently, to generate random samples. The weights here aren't used like that at the moment. Instead, they go into calculations like variances where something would be divided by a weighted sum.

Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:90]: Ok, I think I understand.
So we cannot get any benefit from the fact that all the weights sum up to 1 as in sub-trees it will not be the case anyway. Right?

**Comment by Vika-F** [comment]
@ethanglaser @david-cortes-intel The last question: do we need similar changes in DPC?

**Comment by david-cortes-intel** [comment]
> @ethanglaser @david-cortes-intel The last question: do we need similar changes in DPC?

Last I knew, the DPC versions didn't support sample weights.

**Comment by Vika-F** [comment]
> Last I knew, the DPC versions didn't support sample weights.

I think it supports weights at least in some cases:
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp#L413

**Comment by david-cortes-intel** [comment]
> > Last I knew, the DPC versions didn't support sample weights.
> 
> I think it supports weights at least in some cases: https://github.com/uxlfoundation/oneDAL/blob/main/cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp#L413

@Alexandr-Solovev Could you confirm if weights are supported for random forests on GPU, or is that's just a one-off function?

Currently, they are not enabled in Python:
https://uxlfoundation.github.io/scikit-learn-intelex/2026.1/algorithms.html#id1

**Review by Vika-F — APPROVED** [review]
Don't want to block this PR. Let's make the fix on GPU side later if needed.

**Comment by ethanglaser** [comment]
Sounds good, would also appreciate approval on https://github.com/uxlfoundation/scikit-learn-intelex/pull/3292 as they are related
